packages/@tailwindcss-upgrade/src/template/candidates.ts

```diff
@@ -0,0 +1,149 @@
+  content: string,
+): Promise<{ candidate: Candidate; start: number; end: number }[]> {
+  let scanner = new Scanner({})
+  let result = scanner.getCandidatesWithPositions({ content, extension: 'html' })
+
+  let candidates: { candidate: Candidate; start: number; end: number }[] = []
+  for (let { candidate: rawCandidate, position: start } of result) {
+    for (let candidate of designSystem.parseCandidate(rawCandidate)) {
+      candidates.push({ candidate, start, end: start + rawCandidate.length })
+    }
+  }
+  return candidates
+}
+
+export function printCandidate(candidate: Candidate | null) {
+  if (candidate === null) return 'null'
+  let parts: string[] = []
+
+  for (let variant of candidate.variants) {
+    parts.unshift(printVariant(variant))
+  }
+
+  let base: string = ''
+
+  // Handle negative
+  if (candidate.kind === 'static' || candidate.kind === 'functional') {
+    if (candidate.negative) {
+      base += '-'
+    }
+  }
+
+  // Handle static
+  if (candidate.kind === 'static') {
+    base += candidate.root
+  }
+
+  // Handle functional
+  if (candidate.kind === 'functional') {
+    base += candidate.root
+
+    if (candidate.value) {
+      if (candidate.value.kind === 'arbitrary') {
+        if (candidate.value === null) {
+          base += ''
+        } else if (candidate.value.dataType) {
+          base += `-[${candidate.value.dataType}:${escapeArbitrary(candidate.value.value)}]`
+        } else {
+          base += `-[${escapeArbitrary(candidate.value.value)}]`
+        }
+      } else if (candidate.value.kind === 'named') {
+        base += `-${candidate.value.value}`
+      }
+    }
+  }
+
+  // Handle arbitrary
+  if (candidate.kind === 'arbitrary') {
+    base += `[${candidate.property}:${escapeArbitrary(candidate.value)}]`
+  }
+
+  // Handle modifier
+  if (candidate.kind === 'arbitrary' || candidate.kind === 'functional') {
+    if (candidate.modifier) {
+      if (candidate.modifier.kind === 'arbitrary') {
+        base += `/[${escapeArbitrary(candidate.modifier.value)}]`
+      } else if (candidate.modifier.kind === 'named') {
+        base += `/${candidate.modifier.value}`
+      }
+    }
+  }
+
+  // Handle important
+  if (candidate.important) {
+    base += '!'
+  }
+
+  parts.push(base)
+
+  return parts.join(':')
+}
+
+function printVariant(variant: Variant) {
+  // Handle static variants
+  if (variant.kind === 'static') {
+    return variant.root
+  }
+
+  // Handle arbitrary variants
+  if (variant.kind === 'arbitrary') {
+    return `[${escapeArbitrary(variant.selector)}]`
+  }
+
+  let base: string = ''
+
+  // Handle functional variants
+  if (variant.kind === 'functional') {
+    if (variant.value) {
+      if (variant.value.kind === 'arbitrary') {
+        base += `${variant.root}-[${escapeArbitrary(variant.value.value)}]`
+      } else if (variant.value.kind === 'named') {
+        base += `${variant.root}-${variant.value.value}`
+      }
+    } else {
+      base += variant.root
```

Comment:
You could move this above line 104 and the other conditions won't need to repeat it.